core/common/src/EntityReference.ts

```diff
@@ -0,0 +1,110 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Schema
+ */
+
+import { Id64, Id64String } from "@itwin/core-bentley";
+
+/**
+ * Types of concrete entities. Used for storing strings in JavaScript reference-equality containers which encode
+ * low-level entity information.
+ * @note the values of this enum are unstable, do not depend upon their values between versions of iTwin.js
+ *       (e.g. do not serialize them and load them in another version of iTwin.js and expect them to work)
+ * CodeSpecs are excepted since their JavaScript representation does not derive from [Entity]($backend)
+ * @note the string value of each variant is required/guaranteed to be 1 character
+ * @see EntityReference
+ * @alpha
+ */
+export enum ConcreteEntityTypes {
+  Model = "m",
+  Element = "e",
+  ElementAspect = "a",
+  Relationship = "r",
+}
+
+/**
+ * Adds some utilities to the [[ConcreteEntityTypes]] enum
+ * @alpha
+ */
+export namespace ConcreteEntityTypes {
+  const toBisCoreRootClassFullNameMap = {
+    [ConcreteEntityTypes.Model]: "BisCore:Model",
+    [ConcreteEntityTypes.Element]: "BisCore:Element",
+    [ConcreteEntityTypes.ElementAspect]: "BisCore:ElementAspect",
+    [ConcreteEntityTypes.Relationship]: "BisCore:Relationship",
+  } as const;
+
+  /** used by the transformer to figure out where to check for the existence in a db of a concrete element id
+   * @internal
+   */
+  export function toBisCoreRootClassFullName(type: ConcreteEntityTypes): string {
+    return toBisCoreRootClassFullNameMap[type];
+  }
+}
+
+/**
+ * This id format can be used for storing a unique key for an entity in containers like `Map`.
+ * Elements and non-element entities have different id sequences, they can collide with each other, but not within themselves.
+ * @note for additional utilities that require runtime backend classes, see [EntityReferences]($backend)
+ * @alpha
+ */
+export type EntityReference = `${ConcreteEntityTypes}${Id64String}`;
+
+/** Utility functions for [[EntityReference]] which is a subset of string
+ * @alpha
+ */
+export class EntityReferences {
```

Comment:
Why isn't this `namespace EntityReference`?

Reply:
it was but I thought we preferred static classes so I changed it. Will revert that.

Reply:
it looks like I ended up extending it in core-backend/EntityReference with more backend-only utilities. I think then we should keep this as a class since it is extended

Reply:
`EntityReference` **singular**, so that its declaration is merged with the `type EntityReference` such that all APIs related to that type are grouped together.

Reply:
you're totally right. will fix rn

Reply:
I can't get cross-package namespace merge to work without issues, here is an `MVR` repo https://github.com/MichaelBelousov/typescript-namespace-rexport-mvr

I get errors like this (using the MVR's errors instead of the actual repo but `T` == `EntityReference`):
```sh
│ consumer.ts(7,15): error TS2339: Property 'f' does not exist on type 'typeof T'.
│ consumer.ts(12,6): error TS2709: Cannot use namespace 'T' as a type.
│ consumer.ts(13,6): error TS2709: Cannot use namespace 'T' as a type.
```

I think I will stick with the separate plural name, and see if I can merge the namespaces but the MVR also shows that wasn't working either, but I'll look into it further

Reply:
You are talking about your two `EntityReferences` classes - one in core-common and one in core-backend that extends the other one, right?
That is just plain wacky and is going to confuse everyone.
I'm talking about the type and class (namespace) defined in core-common.

Reply:
I don't care about that. Do you?

Reply:
I will try putting it all in core backend, there was a reason I didnt but I dont remember

Reply:
if I still can't, I will rename the backend namespace to EntityReferenceUtils, which is also wacky since it creates them

Reply:
I believe the reason I moved it was circular references but I still have that issue clearly and am working around it so I'll just put it all in the one package hopefully

Reply:
I can move the namespace but not the type to backend, common's `RenderSchedule.discloseIds` needs to know about `EntityReference` and `EntityReferenceSet`. But I cannot re-export the type from the backend along the namespace without the same compiler errors. So consumers will need to import `EntityReference` twice, once from common for the type and once from backend for the utilities. I think that's at least as confusing. 

I don't see any other options except changing the name of the namespace, I think `EntityReferences` is better than `EntityReferenceUtils`, but feel free to disagree.

Reply:
> consumers will need to import EntityReference twice

Consumers will only have that problem if you insist on using the same name for two different things. Call the thing in core-backend something different.

Reply:
that's the second part of my previous comment. I still prefer `EntityReferences` but feel free to suggest a better name

Reply:
I don't care.